File: js/src/lib/csv-writer.ts
/**
 * Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import csvStringify from 'csv-stringify';
import {stringify} from 'csv-stringify/sync';
import fs from 'fs';
import path from 'path';

import {getLogger} from './logger';
import {IResultWriter, QueryElements, QueryResult} from './types';

export interface CsvWriterOptions {
  destinationFolder?: string | undefined;
  arraySeparator?: string | undefined;
  filePerCustomer?: boolean | undefined;
}
export class CsvWriter implements IResultWriter {
  destination: string | undefined;
  arraySeparator: string;
  filePerCustomer: boolean;
  filename: string | undefined;
  scriptName: string | undefined;;
  appending = false;
  customerRows = 0;
  rowsByCustomer: Record<string, any[][]> = {};
  query: QueryElements | undefined;
  logger;

  constructor(options?: CsvWriterOptions) {
    this.destination = options?.destinationFolder;
    this.arraySeparator = options?.arraySeparator || "|";
    this.filePerCustomer = !!options?.filePerCustomer;
    this.logger = getLogger();
  }

  beginScript(scriptName: string, query: QueryElements) {
    this.appending = false;
    this.query = query;
    this.scriptName = scriptName;

    //let filename = scriptName + ".csv";
    if (this.destination) {
      if (!fs.existsSync(this.destination)) {
        fs.mkdirSync(this.destination, { recursive: true });
      }
      //filename = path.join(this.destination, filename);
    }
    // this.filename = filename;
    // if (fs.existsSync(this.filename)) {
    //   fs.rmSync(this.filename);
    // }
  }

  endScript() {
    //this.filename = undefined;
    this.scriptName = undefined;
  }

  beginCustomer(customerId: string) {
    this.rowsByCustomer[customerId] = [];
  }

  addRow(customerId: string, parsedRow: any[], rawRow: any[]) {
    if (!parsedRow || parsedRow.length == 0) return;
    this.rowsByCustomer[customerId].push(parsedRow);
  }

  _getFileName(customerId: string) {
    let filename = '';
    if (this.filePerCustomer) {
      filename = `${this.scriptName}_${customerId}.csv`;
    } else {
      filename = `${this.scriptName}.csv`;
    }
    if (this.destination) {
      filename = path.join(this.destination, filename);
    }
    return filename;
  }

  endCustomer(customerId: string) {
    let rows = this.rowsByCustomer[customerId];
    if (!rows.length) {
      return;
    }
    let appending = this.appending && !this.filePerCustomer;
    let filename = this._getFileName(customerId);
    let csvOptions: csvStringify.Options = {
      header: !appending,
      quoted: false,
      columns: this.query!.columns.map((col) => col.name),
      cast: {
        boolean: (value: boolean, context: csvStringify.CastingContext) =>
          value ? "true" : "false",
        object: (value: object, context: csvStringify.CastingContext) =>
          Array.isArray(value)
            ? value.join(this.arraySeparator)
            : JSON.stringify(value),
      },
    };
    let csv = stringify(rows, csvOptions);
    fs.writeFileSync(filename!, csv, {
      encoding: "utf-8",
      flag: appending ? "a" : "w",
    });

    if (rows.length > 0) {
      this.logger.info(
        (appending ? "Updated " : "Created ") +
          filename +
          ` with ${rows.length} rows`,
        { customerId: customerId, scriptName: filename }
      );
    }

    this.appending = true;
    this.rowsByCustomer[customerId] = [];
  }
}

export class NullWriter implements IResultWriter {
  beginScript(scriptName: string, query: QueryElements): void|Promise<void> {}
  beginCustomer(customerId: string): void|Promise<void> {}
  addRow(customerId: string, parsedRow: any[], rawRow: any[]): void {}
  endCustomer(customerId: string): void|Promise<void> {}
  endScript(): void|Promise<void> {}
}

File: py/gaarf/io/writers/csv_writer.py
# Copyright 2024 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Module for writing data with CSV."""

# pylint: disable=C0330, g-bad-import-order, g-multiple-import, g-bare-generic

from __future__ import annotations

import csv
import logging
import os
from typing import Literal, Union

import smart_open

from gaarf.io import formatter
from gaarf.io.writers import file_writer
from gaarf.report import GaarfReport


class CsvWriter(file_writer.FileWriter):
  """Writes Gaarf Report to CSV.

  Attributes:
      destination_folder: Destination where CSV files are stored.
      delimiter: CSV delimiter.
      quotechar: CSV writer quotechar.
      quoting: CSV writer quoting.
  """

  def __init__(
    self,
    destination_folder: Union[str, os.PathLike] = os.getcwd(),
    delimiter: str = ',',
    quotechar: str = '"',
    quoting: Literal[0] = csv.QUOTE_MINIMAL,
    **kwargs,
  ) -> None:
    """Initializes CsvWriter based on a destination_folder.

    Args:
      destination_folder: Destination where CSV files are stored.
      delimiter: CSV delimiter.
      quotechar: CSV writer quotechar.
      quoting: CSV writer quoting.
      kwargs: Optional keyword arguments to initialize writer.
    """
    super().__init__(destination_folder=destination_folder, **kwargs)
    self.delimiter = delimiter
    self.quotechar = quotechar
    self.quoting = quoting

  def __str__(self):
    return (
      f'[CSV] - data are saved to {self.destination_folder} '
      'destination_folder.'
    )

  def write(self, report: GaarfReport, destination: str) -> str:
    """Writes Gaarf report to a CSV file.

    Args:
        report: Gaarf report.
        destination: Base file name report should be written to.

    Returns:
        Full path where data are written.
    """
    report = self.format_for_write(report)
    destination = formatter.format_extension(destination, new_extension='.csv')
    self.create_dir()
    logging.debug('Writing %d rows of data to %s', len(report), destination)
    output_path = os.path.join(self.destination_folder, destination)
    with smart_open.open(
      output_path,
      encoding='utf-8',
      mode='w',
    ) as file:
      writer = csv.writer(
        file,
        delimiter=self.delimiter,
        quotechar=self.quotechar,
        quoting=self.quoting,
      )
      writer.writerow(report.column_names)
      writer.writerows(report.results)
    logging.debug('Writing to %s is completed', output_path)
    return f'[CSV] - at {output_path}'
